Home Assistant: add a new package

BKPepe opened the pull request:
Maintainer: me
Compile tested: 
- Turris MOX, cortexa53, OpenWrt master
- Turris Omnia, mvebu, OpenWrt SNAPSHOT r9420-c17a68c

Run tested: 
- Turris MOX, cortexa53, OpenWrt master

Description:

- add: [Home Assistant](https://home-assistant.io)

karlp commented:
linking https://github.com/openwrt/packages/pull/5701/files as related, even though this is the package itself.

karlp commented:
Both the version patches look like things that should go upstream.  At least pyyaml, and the other one should probably got to a >= rather than "just another" fixed point in time?

BKPepe commented:
Yes, I am aware of the pull request and I would like to help @mueslo with his pull request, but it will take me a while as I want to do some things before I can look into it.

____

To be honest, I don't have a clue why some of their dependencies are pinned and others are not. I sent a few pull requests to Home Assistant and all of them were accepted. In the end, it's their decision.
You may look here https://github.com/home-assistant/home-assistant/pull/22997#discussion_r274454049 and I didn't get a reply, why they use it like it.
The same thing is about ruamel.yaml. On their [website](https://bitbucket.org/ruamel/yaml), there is 

> For production systems you should pin the version being used with ``ruamel.yaml<=0.15``.
And it's pinned in Home Assistant.

However, I created 5 days ago issue for PyYAML: https://github.com/home-assistant/home-assistant/issues/23195 , it doesn't seem that it will go through. For attrs, it's a little bit complicated and if I remember correctly some test wasn't passing, but I will try it once again as what I have tested it works.
That's why I didn't send it to devs of Home Assistant yet.

____ 

I would appreciate any feedback regarding it and what you tested as it will help me. :-) Still, we're talking about pre-release version, which I used because of some dependencies as I would need to backport a few commits.

BKPepe commented:
On the other hand, I added a lot of packages for Home Assistant and it would be bad if I didn't show that I have it working. I need to start somehow. ;-)

karlp commented:
You can add more descriptive commentary to the patches perhaps explaining why they are necessary here, and not upstream.  They can be full "git am" patches, so you can have descriptive headers.  See https://github.com/openwrt/packages/blob/master/net/net-snmp/patches/000-cross-compile.patch for example.

I've done no testing whatsoever. I don't use home assistant and don't have python running on any openwrt devices, this is just reviewing the general form :)

jefferyto commented:
ruamel.yaml is changing their API, that's why they are suggest pinning the version for production systems:

> Starting with version 0.15.0 the way YAML files are loaded and dumped is changing. See the API doc for details. Currently existing functionality will throw a warning before being changed/removed. For production systems you should pin the version being used with ``ruamel.yaml<=0.15``.

jefferyto commented:
Regarding the version pinning, I would suggest removing the pin entirely, rather than changing the version numbers. Otherwise, every time a dependency (i.e. the OpenWrt package) is upgraded, you would need to change the patch. (On the other hand, you wouldn't know if an upgraded dependency breaks HA... Hopefully everyone is using semver correctly :laughing:)

feckert commented:
@BkPepe sorry I have set this from draft to pull request!!! This was my fault. I can not undo this :-(

BKPepe commented:
@jefferyto: Yes, it will be a little bit difficult to maintain it. But for now, I am ready to track the changes and maintain it here and in HA.

@feckert: No need to worry about it. :) There's no way to set PR as Draft once it is marked as review.

BKPepe commented:
I've just pushed a commit, which includes a new version 0.92.0 of HA.

Currently, it doesn't work with python3-urllib3.
`pkg_resources.ContextualVersionConflict: (urllib3 1.25.1 (/usr/lib/python3.7/site-packages), Requirement.parse('urllib3<1.25,>=1.21.1'), {'requests'})`
I will need to dig into it a little bit more how to upgrade it.

For now, I have been able to test it with python3-urllib3 with version 1.24.2

Patches changes:
- Added upgrade to ruamel-yaml to 0.15.94
- Added upgrade to pyotp to 2.2.7
- Added upgrade to xmltodict 0.12.0

In all of them, I included a note in their patches about issues or pull request in Home Assistant repository.

BKPepe commented:
A few minutes ago, I pushed another commit to update Home Assistant to 0.92.1.

- Added patch, which upgrades zeroconf to version 0.22.0.
(there is no description in the patch, yet)

Home Assistant tested with Xiaomi gateway with temperature sensors and it works. :-) 
So, the next steps are things in the to-do list, but it's time for that. I will wait for a new version of Home Assistant, which should have fewer patches than now.

BKPepe commented:
Updated it to latest version, which is now 0.93.1.

Changes:

- removed almost all patches (right now, there is only PyYAML, I'd like to get it in next release of hass)

BKPepe commented:
Currently blocked by https://github.com/openwrt/packages/pull/8562

`pkg_resources.DistributionNotFound: The 'importlib-metadata==0.15' distribution was not found and is required by homeassistant`

BKPepe commented on lang/python/homeassistant/Makefile:
```diff
@@ -0,0 +1,78 @@
+#
+# Copyright (C) 2019 CZ.NIC, z. s. p. o. (https://www.nic.cz/)
```
Nope.  Source:
https://openwrt.org/docs/guide-developer/package-policies#copyright_statements
https://github.com/openwrt/packages/blame/master/CONTRIBUTING.md#L30

BKPepe commented on lang/python/homeassistant/Makefile:
```diff
@@ -0,0 +1,78 @@
+	+python3-voluptuous-serialize \
+	+python3-sqlalchemy \
+	+python3-yaml \
+	+python3-zeroconf
```
Fixed.

BKPepe commented on lang/python/homeassistant/files/hass.init:
```diff
@@ -0,0 +1,17 @@
+start_service() {
+        procd_open_instance
+        procd_set_param command "$PROG" -c "$CONF"
+        procd_set_param respawn "${respawn_threshold:-3600}" "${respawn_timeout:-5}" "${respawn_retry:-5}"
```
If I am not mistaken according to documentation for [procd](https://openwrt.org/docs/guide-developer/procd-init-scripts), which is **is still under writing** those values should be as default. I rather put them here, so in the future, I don't need to look at the documentation and search what default values are and if something happens I can slightly edit them.

But I don't mind to remove it. ;)

BKPepe commented on lang/python/homeassistant/patches/0001-Update-dependencies-to-their-latest-version.patch:
```diff
@@ -0,0 +1,81 @@
+-jinja2>=2.10
++bcrypt==3.1.7
++certifi>=2019.06.16
++jinja2>=2.10.1
```
If you are talking about file package_constraints.txt, I can drop my patch for it and see if it works, but I did that when I was doing pull requests to Home Assistant, so I just cherry-picked the commit and included a commit message and I just continue with that. 

About the dependencies in Home Assistant. 
I need to patch them all otherwise it won't work if I had installed an old version of some dependencies. It throws me this error:

> pkg_resources.DistributionNotFound: The 'ruamel.yaml==0.15.97' distribution was not found and is required by homeassistant

However, I'd like to get rid of this patch as much as possible. I'm trying to track changes for dependencies in Hass and here, but the latest version of Hass isn't working for me. I created an issue in their Github repository and in the meantime, I ended up to get to run an old version, which worked for me in the past. I know, it's not perfect right now, but I'd like to finish this PR.

Cynerd replied:
I know but version 2.10.1 is valid for both `jinja2>=2.10` and `jinja2>=2.10.1`.

I am thinking if it would not make sense to just drop whole REQUIRES variable instead of this patching. I would do something like `sed -i '/^REQUIRES = [/,/^]$/{/^ /d};' setup.py`

It is not ideal because there is possibility to miss new packages so it requires manual work as well but I think that it should just work.

Another option is to replace in general all pinning with request on newer version:
`sed -i '/^REQUIRES = [/,/^]$/{s/==/>=/};' setup.py` (this requires more tweaking to also remove upper bounds)

BKPepe replied:
Yes, that's what @miska suggested me yesterday, while I told him that I'm not friend with **sed**, because right now with this solution I need to edit the patch everytime when some dependency changes. I will edit it.

BKPepe commented on lang/python/homeassistant/patches/0001-Update-dependencies-to-their-latest-version.patch:
```diff
@@ -0,0 +1,81 @@
++requests==2.22.0
++ruamel.yaml==0.15.97
+ voluptuous==0.11.5
+ voluptuous-serialize==2.1.0
```
If you are talking about file package_constraints.txt, I can drop my patch for it and see if it works, but I did that when I was doing pull requests to Home Assistant, so I cherry-picked the commit and included a commit message and I just continue with that.

BKPepe commented on lang/python/homeassistant/files/hass.init:
```diff
@@ -0,0 +1,17 @@
+STOP=10
+
+PROG=/usr/bin/hass
+CONF=/srv/.homeassistant
```
I'm glad that you asked. I would like to see it there as well.
Unfortunately, in the config folder, there isn't only one configuration file [configuration.yaml](https://www.home-assistant.io/docs/configuration/), but together 9 files and 3 directories. 

For example:
**Files:**
- database (home-assistant_v2.db), in my small setup it writes 2 MB / per day.

**Folders**
- .storage (files there together has 12,9 KB)
- deps
If you don't have some dependencies for Hass and during the startup, it put them here
`/srv/homeassistant/deps/lib/python3.7/site-packages`
In my case, there are PyQRCode, mutagen, home_assistant_frontend and so on. When the Hass managed those dependencies itself, it's fewer things to maintain.

About the location /srv. It's a question. In one case, you shouldn't do excessive writes on the internal storage and on the other hand, you don't want to lose all the data during reboot or power off. From my point of view, I can write to description that Hass has a database and if you want to install it and run it with the default configuration, which I provided you need to mount a flash drive to /srv, but who will read the description? If I put /var, I don't see it very user-friendly as users will need to edit the configuration path in CLI also I would need to patch it in Turris feeds.

BKPepe commented:
Squashed commits, rebased on current master, this one is ready for review unfortunately with an older version of Home Assistant, which I explained in comments and it is also included in the commit message of one of my patches.

@jefferyto , @commodo  would you please look at it? :)

commodo commented on lang/python/homeassistant/Makefile:
```diff
@@ -0,0 +1,74 @@
+
+define Py3Package/$(PKG_NAME)/install
+	$(INSTALL_DIR) $(1)/usr/bin
+	$(CP) $(PKG_INSTALL_DIR)/usr/bin/* $(1)/usr/bin/
```
curios: won't `/usr/bin` stuff be installed automatically?
there is some auto-magic rules that should do this; i am not recalling this 100%; but it's also fine to leave this

BKPepe replied:
They are when `install` section is not defined.

If I define `install` section like this
```
define Py3Package/$(PKG_NAME)/install
	$(INSTALL_DIR) $(1)/etc/init.d
	$(INSTALL_BIN) ./files/hass.init $(1)/etc/init.d/hass
endef
```

Then in the package are
- site-packages
- init script

Nothing more.

commodo replied:
oh right;
the default one does that, but you are overriding it;

thanks & sorry for the noise;
i think i did not have any coffee before i asked

BKPepe replied:
That's fine. You don't need to worry about it.

commodo commented on lang/python/homeassistant/patches/0001-Update-dependencies-to-their-latest-version.patch:
```diff
@@ -0,0 +1,81 @@
+ pytz>=2019.01
+-pyyaml>=3.13,<4
+-requests==2.21.0
+-ruamel.yaml==0.15.94
```
we may want to find a better mechanism here than patching it like this;
i have no idea for one atm;
but we'll see

BKPepe replied:
I wouldn't say that it is so difficult as packages in stable releases (OpenWrt 18.06 and 19.07) are not updated so often.

Right now, there are two solutions

1. Keep tracking changes here and in Hass
2. Remove the entire pinning which was suggested by @jefferyto

commodo replied:
no strong preference from my side;
a proper mechanism [that works for all/most cases] would be a bit of work anyway [at this point in time];

it's not one of those battles that's worth picking right now;
there are more important things to do until that;

commodo commented on lang/python/homeassistant/files/hass.init:
```diff
@@ -0,0 +1,17 @@
+
+start_service() {
+        procd_open_instance
+        procd_set_param command "$PROG" -c "$CONF"
```
if you run this with `procd_set_param command "$PROG" -c "$CONF" --daemon`  would it work to remove 
```
procd_set_param stdout 1
procd_set_param stderr 1`
```
?

Cynerd replied:
Why do you want to remove standard output redirection to syslog? I think that it is a good approach. I don't see reason why we should avoid it.

Also parameters you posted are not good idea as well. Using `--daemon` with procd service is bad. Procd looses track of service if it does double fork (which is what `--daemon` instructs it to do). This means that there is effectively no way to kill service trough procd. It is something I encountered with syslog-ng recently so I am clearly aware that it is not a good idea.

commodo replied:
oh right;
i forgot that procd requires services in foreground [been a while since i used it more thoroughly]
the `--daemon` arg is stupid to use now; so, my bad

regarding:
> Why do you want to remove standard output redirection to syslog? I think that it is a good approach. I don't see reason why we should avoid it.

things may have changed in a while;
but, personally, i liked it when services would use syslog directly, so that procd would not need to do any output redirection to syslog;

i sometimes found that procd's output redirection was sometimes buggy [but that probably isn't the case anymore];

some services, do allow the program to stay in foreground, while output-ing directly to syslog
[ i was looking for this when i suggested `--daemon` ]

neheb replied:
I'd use procd_append_param. Although that's just a style thing.

commodo commented on lang/python/homeassistant/Makefile:
```diff
@@ -0,0 +1,78 @@
+
+	$(INSTALL_DIR) $(1)/etc/init.d
+	$(INSTALL_BIN) ./files/hass.init $(1)/etc/init.d/hass
+endef
```
IIRC: `filespec` is typically used for packages that go under `/usr/lib/pythonX.Y/`

jefferyto commented on lang/python/homeassistant/Makefile:
```diff
@@ -0,0 +1,74 @@
+    +python3-voluptuous-serialize \
+    +python3-sqlalchemy \
+    +python3-yaml \
+    +python3-zeroconf
```
Comparing the list of dependencies here and the ones in [setup.py](https://github.com/home-assistant/home-assistant/blob/0.93.1/setup.py#L34-L53), it looks like you added a few (aiohttp-cors, ifaddr, netdisco, pyotp, sqlalchemy, zeroconf) that are required by [different components](https://github.com/home-assistant/home-assistant/blob/0.93.1/requirements_all.txt) but not core? (Actually I couldn't find any direct requirements for ifaddr.)

If this is the case, I think it would be better to require only core's dependencies and let users decide which components they want to use and then install those component's dependencies individually?

jefferyto commented:
A further thought on components (though I know nothing about Home Assistant and haven't really looked into it yet): If there are a large number of optional components (I would guess that different users will have different hardware they want to control), perhaps it would help to have separate packages for each component (or related components, again I don't know the specifics)? This would keep the core package smaller, and component packages can have their own dependencies. (You could also have meta-packages that installs all components, a recommended minimum set, or for different setups, etc.)

jefferyto commented:
Also a wild guess for home-assistant/home-assistant#24982: In 0.94.0 they added importlib-metadata as a new dependency (home-assistant/home-assistant#24114), not sure if you were aware.

I have no idea if the error messages you got were because this package was missing, but I thought I would mention it :laughing:

TimelessNL commented:
I'm trying your Hass package but it seems that there are some "core" deps missing? During Hass's initial startup it tries to install 'PyQRCode 1.6.1', which of course fails on target. But after building that package on my host and installing it on target, Hass then starts to install a bunch of other packages to load the frond-end:
```
Sun Aug 11 21:06:21 2019 daemon.err hass[1939]:   Could not find a version that satisfies the requirement distro==1.4.0 (from versions: )
Sun Aug 11 21:07:19 2019 daemon.err hass[1939]:   Could not find a version that satisfies the requirement mutagen==1.42.0 (from versions: )
Sun Aug 11 21:08:15 2019 daemon.err hass[1939]:   Could not find a version that satisfies the requirement xmltodict==0.12.0 (from versions: )
Sun Aug 11 21:09:10 2019 daemon.err hass[1939]:   Could not find a version that satisfies the requirement hass-nabucasa==0.12 (from versions: )
Sun Aug 11 21:10:06 2019 daemon.err hass[1939]:   Could not find a version that satisfies the requirement zeroconf==0.22.0 (from versions: )
Sun Aug 11 21:11:01 2019 daemon.err hass[1939]:   Could not find a version that satisfies the requirement PyNaCl==1.3.0 (from versions: )
Sun Aug 11 21:11:57 2019 daemon.err hass[1939]:   Could not find a version that satisfies the requirement home-assistant-frontend==20190514.0 (from versions: )
```
After that Hass seems to be ready but the homepages shows:
`404: Not Found`
I never really checked all the python libaries Hass used on my Ubuntu installation. But I was shocked to see how many python packages Hass has listed. How are we ever going to keep up with that?

BKPepe commented:
It shows you `404 Not Found`, because it was not able to install home-assistant-frontend==20190514.0 and it was usually of missing dependencies, which pip for some reason doesn't install most likely due to error (run out of storage space or RAM space). You should look also to pip log. A few times I find there something useful.

In Makefile, right now, they are dependencies, which are must-have otherwise hass won't run. When haas is starting it will install additional dependencies itself and it is managed by hass, so fewer things to maintain. What I have observed that not many dependencies are changing so often, so it wouldn't be hard to keep up with that. 

zeroconf, xmltodict are already in OpenWrt, so it's a little bit weird, why it couldn't find them. PyNaCL is being worked on.

BKPepe commented:
I was running it on a device, which had 512 MB, but it was running just fine until it stopped working on newer versions due to unknown reasons for me. However, they have listed as a minimum requirement 1 GB RAM. In upcoming days, I should have a device with 2 GB RAM, so I will try to finish the PR and update hass to its latest version

TimelessNL commented:
Yeah it strange that it does not find zeroconfig, it is installed:
```
root@OpenWrt:~# opkg list | grep zero
python3-zeroconf - 0.23.0-1
```
I'm running it on a PC Engines ALIX (GEODE/256MB RAM)
so RAM is on a tight budget here, I never expected hass to install any more deps. So I did not connect it to the internet yet. That's probably why it fails to install those deps.

But one question though, what if those python libaries have native code in them, then gcc and such is needed on target. That's not very "OpenWrt" like.

BKPepe commented:
This time updated to version 0.97.2. 

- Blocked by #8562 as you need python3-importlib-metadata (which depends on zipp).
- There is just one patch, which updates python-ruamel to the latest version of 0.15. It was sent to Home Assistant. Will be removed soon.

When you run it for the first time, you will need to startup Hass and once it homeassistant is initialized, you need to restart it and then frontend will be shown.

The issue mentioned with zeroconf is fixed.

stintel commented on lang/python/homeassistant/Makefile:
```diff
@@ -0,0 +1,77 @@
+    +python3-voluptuous-serialize \
+    +python3-sqlalchemy \
+    +python3-yaml \
+    +python3-zeroconf
```
Can you keep the dependencies in alphabetical order?

BKPepe replied:
Yes. Should be done.

stintel replied:
Not entirely. python3-r* should go before python3-s*

BKPepe replied:
I see. Right. My fault.

BKPepe replied:
Now, should be fixed.

stintel commented on lang/python/homeassistant/Makefile:
```diff
@@ -0,0 +1,77 @@
+define Package/$(PKG_NAME)
+  SUBMENU:=Python
+  SECTION:=lang
+  CATEGORY:=Languages
```
I would rather put this in utils/Utilities like Domoticz.

BKPepe replied:
Ok, I dont have any issues with that.

jefferyto commented:
> In Makefile, right now, they are dependencies, which are must-have otherwise hass won't run. When haas is starting it will install additional dependencies itself and it is managed by hass, so fewer things to maintain.

(Apologies for missing this discussion earlier.) I'm not sure I like the idea of software calling pip directly to install their own dependencies at run-time. This happens without the user's knowledge, and as @TimelessNL mentioned, some packages may require a dev toolchain that may not exist on the device.

BKPepe commented:
> (Apologies for missing this discussion earlier.)

That's OK. This is still blocked by https://github.com/openwrt/packages/pull/10443 (and that's why I didn't look further to resolve discussions above) and unfortunately https://github.com/home-assistant/home-assistant/issues/31021 , where is a proof that I'm running it, but I need to install importlib-metadata even when it is part of Python 3.8.  

> I'm not sure I like the idea of software calling pip directly to install their own dependencies at run-time.

I am not aware if there is a different (better?) solution then using pip to install it other dependencies. If you are going to take a look at https://github.com/home-assistant/home-assistant/blob/dev/requirements_all.txt there are a lot of packages, which should be packaged to OpenWrt. Each user has its own configuration/setup/usage. As an example, I'm using Xiaomi sensors, which require PyXiaomiGateway. It should not be a dependency in Makefile of Home Assistant. But I don't like an approach that users, who would like to have Home Assistant on their router, they should install all dependencies manually. It's a lot of stuff to take care of it.

Here you go gist: https://gist.github.com/BKPepe/d67c82996389f68fe63b67ec87cc8a4e to see which packages Home Assistant installed on its own.  There are some duplicates, which were caused as I was upgrading from the older version.

BKPepe commented:
I should require to compile Home Assistant just on mvebu, aarch64 and x84_x64. Other platforms won't be most likely supported due to HW limitation as haas require at least 1 GB RAM. It runs potentially with 512 MB, but that's not supported as some unexpected things may happen.

aparcar commented:
Could you please update this to the latest HA release? Also I'm fine this is only compiled for a limited number of targets.

BKPepe commented:
Thanks for pinging me, @aparcar. Yeah, I will try to update it during this week. I need to find my gateway first to be sure that it works as before.

aparcar commented:
Friendly ping

BKPepe commented:
I started looking at it since yesterday. I can not use the version which was released yesterday (2021.1.1) as I have some issues with that. Version, which I pushed seems to be working at least for me, but it requires an additional one package, which is missing in OpenWrt but that package requires additional 2-3 dependencies. I packaged it and will send a PR within a few hours. They can be installed via pip if you want to test it right now.

juanesf commented:
> Debería requerir compilar Home Assistant solo en mvebu, aarch64 y x84_x64. Es muy probable que otras plataformas no sean compatibles debido a la limitación de HW, ya que haas requiere al menos 1 GB de RAM. Se ejecuta potencialmente con 512 MB, pero eso no es compatible ya que pueden suceder algunas cosas inesperadas.



> I should require to compile Home Assistant just on mvebu, aarch64 and x84_x64. Other platforms won't be most likely supported due to HW limitation as haas require at least 1 GB RAM. It runs potentially with 512 MB, but that's not supported as some unexpected things may happen.

most allwinner (sunxi) boards have at least 512 ram and several cores although they are arm, I'm sure it runs on bpi-m2-ultra

wulfy23 commented:
not sure if suggested... too many comments to catch up on...: utilities > automation?

BKPepe commented:
Unfortunately, because of https://www.home-assistant.io/blog/2025/05/22/deprecating-core-and-supervised-installation-methods-and-32-bit-systems/ I need to close this pull request as it is not going to be supported anymore.

aparcar commented:
I'm still interested in HA on OpenWrt, specifically the One since it supports Zigbee via the MicroBus. Should we look into getting those containers ready with APK? We tried that before and Debian would work fine in an APK container.

@dangowrt wwdd?
